src/main/java/org/basseur/adventofcode/advent2018/days/day09/CircleDeque.java

```diff
@@ -0,0 +1,58 @@
+package org.basseur.adventofcode.advent2018.days.day09;
+
+import java.util.ArrayDeque;
+import java.util.Collection;
+
+/**
+ * Implementation of a circular ArrayDeque which adds a method to rotate the Deque.
+ *
+ * @param <E> the type of elements held in this collection
+ * @author Jan Philipp G&ouml;lz
+ */
+public class CircleDeque<E> extends ArrayDeque<E> {
```

Comment:
nice idea, I didn't know the ArrayDeque before :)

Reply:
How is the performance here? We used linked lists to solve this...Matthias implementation was rather slow (2-3 h), mine a little faster (< 10 min) but with a improved implementation - and yours?

Does the user have to wait quite some time to see the result? Now things get interesting. We should consider to come back to the user via email or so, if the calculation takes too long...This is why were thinking about persisting results and user accounts, etc :)

Reply:
I wanted to implement this as a circle and found information about doubly linked lists and circular doubly linked lists. Somewhere I found, that you could do that with Deque (pronounced deck), which is short for _double-ended queue_.

And it is extremely fast: For part 1 it takes about 1.8 ms, for part 2 it takes about 284.2 ms (on my machine)! :D
Try here: https://advent-of-code-app-pr-73.herokuapp.com/